### PR TITLE
Blog Article Page Layout

### LINKED ISSUE #135: Blog article page layout
This is the page that renders the HTML from the editor.
https://www.figma.com/file/k6TNqd4V49TsJlsvaZKnQV/International-Student-Mentorship-Program?node-id=303%3A469

### PR DESCRIPTION
**Pull request by vickyadebona**
Closes #135

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -39,7 +39,7 @@ services:
     stdin_open: true
     tty: true
     environment:
-      REACT_APP_API_HOST: "http://localhost:8000"
+      REACT_APP_API_HOST: "http://18.144.82.161:8000"
```
Comment (douglaschan32167) [inline docker-compose.yml:42]: Can you change this back before submitting?

Comment (vickyadebona) [reply docker-compose.yml]: 👍

File: frontend/src/components/RenderBlog/RenderBlog.jsx [hunk 2]
```diff
@@ -1,14 +1,17 @@
...
+const RenderBlog = ({ initial_content }) => {
+  return (
+    <BlogContainer>
+      {ReactHtmlParser(initial_content)}
+    </BlogContainer>
+  )
 };
 
-export default RenderBlog;
+export default RenderBlog;
```
Comment (dbtung) [inline frontend/src/components/RenderBlog/RenderBlog.jsx:17]: nit: End your files with a newline

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 3]
```diff
@@ -1,17 +1,246 @@
...
 import { requests } from 'utils/agent';
+import moment from 'moment';
 import fallbackBlogpostContentData from 'fallback_data/blogpostcontent.json';
+import { Embed } from 'semantic-ui-react';
 
+import Styled from 'styled-components';
+import sizes from '../../styles/sizes';
 import RenderBlog from 'components/RenderBlog';
 
+const arrowLeft = require('../../images/arrow-left-purple.png');
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:12]: Can this be `import arrowLeft from '../../images/arrow-left-purple.png`?

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 4]
```diff
@@ -1,17 +1,246 @@
...
 class Blogpost extends Component {
   constructor(props) {
     super(props);
     this.state = {
-      content:
-        'This is the blogpost page for blogpostcontent id ' +
-        props.match.params.id,
+      content: props.match.params.id,
       blogpostcontent_id: parseInt(props.match.params.id, 10),
       blogpostcontent: {}
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:245]: Few thoughts:
- What's the different between content / blogpostcontent_id? From the default values, seems like they are the same, but I'm assuming they are different. So maybe pick a more appropriate default value for content?
- Should have a default value for `blogPostData`
- Can we change the naming here to all be camelCase?

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 5]
```diff
@@ -21,16 +250,17 @@ class Blogpost extends Component {
...
             });
-          });
       } catch (error) {
-        console.log('there was an error in blogposts');
-        console.log(fallbackBlogpostContentData[0]['id']);
+        console.log("Could not load blogpost: ", error);
+        // Use fallback blog data if we get no response from the API
         let fallbackBlogpostContent = fallbackBlogpostContentData.filter(
           blogpostJson => {
             return this.state.blogpostcontent_id === blogpostJson.id;
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:266]: If we're going to use fallback data, seems like we should just return something, even if the IDs don't match.

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 6]
```diff
@@ -49,12 +279,86 @@ class Blogpost extends Component {
...
+      headerMedia: ''
+    }
+  }
+
+  parseVideoLink(videoLink) {
+    // Video ID is used in Embed component to determine 
+    // where to find the video link on YouTube
+    const videoId = videoLink.includes('embed/') 
+                ? videoLink.split('embed/')[1]
+                : videoLink.split('v=')[1];
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:302]: Can we change the indentation here to match what you have above (two spaces)?

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 7]
```diff
@@ -49,12 +279,86 @@ class Blogpost extends Component {
...
+    if(videoId.includes('&')) {
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:303]: nit: space after `if` - here and below

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 8]
```diff
@@ -49,12 +279,86 @@ class Blogpost extends Component {
...
+      const videoId = this.parseVideoLink(headerMedia);
+      return (
+        <HeaderVideo 
+          id={videoId}
+          source='youtube'
+          iframe={{
+            allowFullscreen: true
+          }}
+        >
+        </HeaderVideo>
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:322]: nit: Since there's no children, don't need separate open/closing tags.

File: frontend/src/pages/Blogs/Blogpost.jsx (hunk 4 above)
Comment (douglaschan32167) [reply frontend/src/pages/Blogs/Blogpost.jsx:245]: blogpostcontent_id is just a number (initially read from url)
content will be the entire json object (title, body, etc)

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 9]
```diff
@@ -49,12 +279,86 @@ class Blogpost extends Component {
...
   render() {
+    const { bio, topic, headerMedia } = this.getBlogData();
     return (
-      <RenderBlog
-        blogpostcontent_id={this.state.blogpostcontent_id}
-        initial_content={this.state.content}
-      />
+      <Container>
+        <BlogContentContainer>
+          <BlogHomeLink to="/blog">
```
Comment (douglaschan32167) [inline frontend/src/pages/Blogs/Blogpost.jsx:334]: can we make this link to "/blog-list" for now since we're cutting the blog landing page from the v1?

File: frontend/src/pages/Blogs/Blogpost.jsx (hunk 5 above)
Comment (douglaschan32167) [reply frontend/src/pages/Blogs/Blogpost.jsx:266]: I kind of like the way it works now where if you type in an invalid blogpostcontent id into the url (like http://localhost:3000/blogpost/99) then you see that it says "the requested blogpost was not found." I think if it showed a blogpost with the wrong id it would be confusing, but I'm not that set in this.

Other thoughts:
-should we still render author, date, and media if the blogpost isn't found?
-maybe later we could add a json object that defines a "blogpost not found blogpost" with a special image and message that tells the user that there's no blogpostcontent with that id (kind of like a specialized 404 page). Don't worry about it for the v1 though.

File: frontend/src/pages/Blogs/Blogpost.jsx (hunk 4 above)
Comment (douglaschan32167) [reply frontend/src/pages/Blogs/Blogpost.jsx:245]: sorry, I mean content is the body_content field of blogpostData. That's the string that the html renderer will render.

My vision for fallback data was so that we would have at least a few blogposts available even if the backend dies. If we return a default blogpost for invalid ids, though, then I think it should either be a blogpost that communicates that this is not a valid id. I think if we return an actual content blogpost, we run into the situation where someone sees the blogpost, links it to his friend, then later someone writes another blogpost so the id becomes valid, and now the url points to something else.

File: frontend/src/pages/Blogs/Blogpost.jsx (hunk 5 above)
Comment (dbtung) [reply frontend/src/pages/Blogs/Blogpost.jsx]: Yeah, either option - I think as long as it's clear to the user that something there's a problem.

File: frontend/src/pages/Blogs/Blogpost.jsx (hunk 4 above)
Comment (dbtung) [reply frontend/src/pages/Blogs/Blogpost.jsx]: Got it - in that case, I think an appropriate default value for `content` could be the empty string.

**Comment by dbtung** [comment]
@vickyadebona Is this ready for another look?

File: frontend/src/components/RenderBlog/RenderBlog.jsx [hunk 10]
```diff
@@ -1,14 +1,17 @@
 import React from 'react';
 import ReactHtmlParser from 'react-html-parser';
-import styled from 'styled-components';
+import Styled from 'styled-components';
```
Comment (albertjke) [inline frontend/src/components/RenderBlog/RenderBlog.jsx:3]: NIT: Could we use lowercase? This is the convention and enables linting

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 11]
```diff
@@ -1,60 +1,402 @@
...
-              content: result.body_content
+        .get('blogpostcontent/' + this.state.blogPostContentId + '/')
+            .then(result => {
+              this.setState({
+                blogPostContent: result,
+                content: result.body_content,
+                blogPostData: result.blogpost,
+                blogPostExists: true
+              });
             });
```
Comment (albertjke) [inline frontend/src/pages/Blogs/Blogpost.jsx:285]: Could we put this under Blogposts in agent.js? Would be nice to consolidate all urls in one place

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 12]
```diff
@@ -1,60 +1,402 @@
...
+`;
+
+const BlogContentContainer = Styled.div`
+  display: flex;
+  flex-direction: column;
+  width: 100%;
+
+  @media (max-width: ${sizes.tablet}px) {
+    padding: 0 5%;
+  }
```
Comment (albertjke) [inline frontend/src/pages/Blogs/Blogpost.jsx:26]: NIT: Can use  ```${media.tablet`style: value;`}``` here

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 13]
```diff
@@ -1,60 +1,402 @@
...
+`;
+
+const HeaderVideo = Styled(Embed)`
+  width: 1050px;
+  height: 590px;
+  align-self: center;
+  margin-bottom: 48px;
+`;
+
+const BlogDataContainer = Styled.div`
```
Comment (albertjke) [inline frontend/src/pages/Blogs/Blogpost.jsx:134]: A lot of these header styles are already defined in globalStyles (add `serif` class to get PTSerif)

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 14]
```diff
@@ -1,60 +1,402 @@
...
+  flex-direction: column;
+  width: 100%;
+
+  @media (max-width: ${sizes.tablet}px) {
+    padding: 0 5%;
+  }
+`;
+
+const BlogTopic = Styled.div`
+  font-family: ${props => props.theme.fonts.Poppins};
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:30]: Just wondering ... is there a reason why we can't just `theme` and reference `${theme.fonts.Poppins}` here?

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 15]
```diff
@@ -1,60 +1,402 @@
...
+        <GetConnected>
+          Get connected with a mentor today
+          <ApplyNowBtn to="/application-form">APPLY NOW</ApplyNowBtn>
+        </GetConnected>
+        { /* Leaving below commented out until blog cards are ready */ }
+        {/* <RecommendedArticlesHeader>
+            <Line></Line>
+            Recommended Articles
+            <Line></Line>
+        </RecommendedArticlesHeader> */}
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:398]: Since this component is already pretty chunky, maybe it would be good to factor this part out to a separate component. Since it's going to have it's own API calls as well.

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 16]
```diff
@@ -1,60 +1,402 @@
...
+          { this.renderHeaderMedia(headerMedia || '') }
+          <BlogDataContainer>
+            <RenderBlog
+              initialContent={this.state.content}
+            />
+          </BlogDataContainer>
+        </BlogContentContainer>
+        <GetConnected>
+          Get connected with a mentor today
+          <ApplyNowBtn to="/application-form">APPLY NOW</ApplyNowBtn>
```
Comment (douglaschan32167) [inline frontend/src/pages/Blogs/Blogpost.jsx:391]: The url for the application form was actually changed to /apply.

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 17]
```diff
@@ -1,60 +1,402 @@
...
   render() {
+    if (!this.state.blogPostExists) {
+      return (
+        <ErrorPageContainer>
+          <ErrorMessage>The requested blog was not found.</ErrorMessage>
+          <BlogHomeErrorLink to="/blog-list">
+            <ArrowErrorIcon src={arrowLeft} alt="arrow-left"/>
+            Blog Home
+          </BlogHomeErrorLink>
+        </ErrorPageContainer>
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:366]: If the blog post does not exist, do we still want to include the GetConnected and RecommendedArticles sections?

**Comment by douglaschan32167** [comment]
For some reason my header media won't load
image

File: frontend/src/pages/Blogs/Blogpost.jsx (hunk 14 above)
Comment (vickyadebona) [reply frontend/src/pages/Blogs/Blogpost.jsx:28]: the theme can be accessed through props since the whole app is wrapped in a context provider that has the theme file as a prop. This is the standard practice I've seen on other applications that use styled components, but also doing it this way allows for one less thing to import.

Comment (dbtung) [reply frontend/src/pages/Blogs/Blogpost.jsx:28]: Hmmm, OK - I would actually recommend importing the `theme` and just referencing it directly - since it's one less level of indirection, and also because that's what we're doing throughout the rest of our app here.

File: frontend/src/pages/Blogs/Blogpost.jsx (hunk 17 above)
Comment (vickyadebona) [reply frontend/src/pages/Blogs/Blogpost.jsx:372]: I've added a loading state and a 404-like page where those things won't render if the blog post doesn't exist.

**Comment by vickyadebona** [comment]
@dbtung @albertjke ready for review again

File: frontend/src/pages/Blogs/Blogpost.jsx (hunk 13 above)
Comment (vickyadebona) [reply frontend/src/pages/Blogs/Blogpost.jsx]: These styles are specifically for the html that renders the blog post. I added them since the global styles don't seem to be hitting this particular section (unless I'm supposed to inject is somehow? Not sure how to go about that).

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 18]
```diff
@@ -1,61 +1,404 @@
...
     const populateBlogpostData = async () => {
       try {
-        await requests
-          .get('blogpostcontent/' + this.state.blogpostcontent_id + '/')
-          .then(result => {
-            this.setState({
-              blogpostcontent: result,
-              content: result.body_content
+        await BlogPost
+        .get(this.state.blogPostContentId)
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:273]: nit: indent +2

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 19]
```diff
@@ -1,61 +1,404 @@
...
           });
         }
       }
     };
     populateBlogpostData();
   }
 
+  getBlogData() {
+    // Making sure associated blog information is defined
+    return (this.state.blogPostData && this.state.blogPostData !== {}) ? {
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:312]: Could just shorten this to `this.state.blogPostData !== {}`

File: frontend/src/pages/Blogs/Blogpost.jsx (hunk 17 above)
Comment (dbtung) [reply frontend/src/pages/Blogs/Blogpost.jsx:372]: Yeah, I guess my point is whether or not we still want to render the GetConnected / RecommendedArticles sections even on that 404-like page. Or even when things are loading, since that content should be independent of whether or not the blog data is loading.

File: frontend/src/pages/Blogs/Blogpost.jsx [hunk 20]
```diff
@@ -1,61 +1,404 @@
...
+          <BlogContentContainer>
+            <BlogHomeLink to="/blog-list">
+              <ArrowLeft src={arrowLeft} alt="arrow-left"/>
+              Blog Home
+            </BlogHomeLink>
+            <BlogTopic>{topic || this.state.blogPostContent.title_content}</BlogTopic>
+            <BlogTitle>{this.state.blogPostContent.title_content}</BlogTitle>
+            <DateAndAuthor>
+              {moment(this.state.blogPostContent.publish_at).format('MMMM DD, YYYY')} by {bio ? bio : 'Unknown Author'}
+            </DateAndAuthor>
```
Comment (dbtung) [inline frontend/src/pages/Blogs/Blogpost.jsx:387]: Looking at this, when there's no author, I think it's better just to leave off the by Unknown Author - seems a little strange to call attention to the fact that we don't know who is writing our blog posts.

**Comment by vickyadebona** [comment]
@douglaschan32167 @albertjke @dbtung I made the last PR changes, is this ready to merge?

**Comment by douglaschan32167** [comment]
Just made one minor change that makes Jesse's bloglist page link to your blogposts. Otherwise it's good to go
